### DISCUSSION TITLE
nixpkgs/overriding

### DISCUSSION THREAD

**Discussion by giscus** [ryan4yin/nixos-and-flakes-book#182] [Announcements]
# nixpkgs/overriding

An unofficial and opinionated book for beginners

https://nixos-and-flakes.thiscute.world/nixpkgs/overriding

**Comment by RobotDisco** [comment]
It's not clear to me why we start with talking about the `override` function and then talk about the `overrideAttrs` function. Do we need to explicitly check to know which function is needed for a given package?

**Reply by dibenzepin** [reply]
the way i see it, `override` is used for the attributes passed into the function (the bits at the very top), while `overrideAttrs` is for the bits in the actual definition...you can read more [here](https://ryantm.github.io/nixpkgs/using/overrides/).

so, in the example above, `pkgs.hello` is defined as:
```nix
{ callPackage
, lib
, stdenv
, fetchurl
, nixos
, testers
, hello
}:

stdenv.mkDerivation (finalAttrs: {
  pname = "hello";
  version = "2.12.1";

  src = fetchurl {
    url = "mirror://gnu/hello/hello-${finalAttrs.version}.tar.gz";
    sha256 = "sha256-jZkUKv2SV28wsM18tCqNxoCZmLxdYH2Idh9RLibH2yA=";
  };

  doCheck = true;

  # ...
})
```

now if you wanted to change the `src`, you'd use `overrideAttrs`, but if you wanted to change, say the `tester` arg, you'd use `override`.

a useful use of this is with the [sudo](https://github.com/NixOS/nixpkgs/blob/nixos-unstable/pkgs/by-name/su/sudo/package.nix) package:
```nix
{
  lib,
  stdenv,
  fetchurl,
  buildPackages,
  coreutils,
  pam,
  groff,
  sssd,
  nixosTests,
  sendmailPath ? "/run/wrappers/bin/sendmail",
  withInsults ? false,
  withSssd ? false,
}:

stdenv.mkDerivation (finalAttrs: {
  pname = "sudo";
  # be sure to check if nixos/modules/security/sudo.nix needs updating when bumping
  # e.g. links to man pages, value constraints etc.
  version = "1.9.16p2";

  __structuredAttrs = true;

  src = fetchurl {
    url = "https://www.sudo.ws/dist/sudo-${finalAttrs.version}.tar.gz";
    hash = "sha256-l2qlbT47KnVZMweGQoit23SMnBNuJdlanMaZqvp3I5w=";
  };
  ...
}
```
...the `withInsults` arg (which defaults to `false`) can be overridden like so:
```nix
security.sudo.package = pkgs.sudo.override { withInsults = true; };
```
